### PR TITLE
Allow `kustomize localize` to process directories

### PR DESCRIPTION
**Pull request by annasong20**
This PR adds directory processing to `kustomize localize`. Certain remote-handling code introduces additional complexity, and so will be implemented in a separate PR. The directory localization logic is tested on the `Components` field.

### CODE REVIEW

File: api/internal/localizer/localizer_test.go [hunk 1]
```diff
@@ -71,13 +74,57 @@ func createLocalizer(t *testing.T, fSys filesys.FileSystem, target string, scope
...
+		visited[path] = struct{}{}
+
+		if _, exists := visited[path]; !exists {
+			t.Errorf("expected path %q not found", path)
+		}
+		return nil
+	})
+	require.NoError(t, err)
+}
+
```
Comment (annasong20) [inline api/internal/localizer/localizer_test.go:120]: Same logic as introduced in #4865

File: api/internal/localizer/util.go [hunk 2]
```diff
@@ -99,3 +105,12 @@ func hasRef(repoURL string) bool {
 	}
 	return repoSpec.Ref != ""
 }
+
+// locRootPath returns the relative localized path of the validated root url rootURL, where the local copy of its repo
+// is at repoDir and the copy of its root is at rootDir
+// TODO(annasong): implement
+func locRootPath(rootURL string, repoDir filesys.ConfirmedDir, rootDir filesys.ConfirmedDir) string {
+	_ = rootURL
+	_, _ = repoDir, rootDir
```
Comment (annasong20) [inline api/internal/localizer/util.go:114]: These statements allow named arguments for use in the spec.

File: api/internal/localizer/localizer.go [hunk 3]
```diff
@@ -75,9 +78,55 @@ func (lc *Localizer) Localize() error {
...
+	defer func() { _ = ldr.Cleanup() }()
+
+	var locPath string
+	if repo := ldr.Repo(); repo != "" {
+		// TODO(annasong): check if able to add localize directory
+		locPath = locRootPath(path, filesys.ConfirmedDir(repo), filesys.ConfirmedDir(ldr.Root()))
+	} else {
+		locPath, err = filepath.Rel(lc.ldr.Root(), ldr.Root())
+		if err != nil {
+			//nolint:gocritic // should never occur, but if hit, something fundamentally wrong; should immediately exit
```
Comment (natasha41575) [inline api/internal/localizer/localizer.go:111]: When your comments have multiple parts with different subjects, could you please make them complete sentences? It is hard to read and requires more effort for the reader to parse as it is. 

```
This should never occur; if it does, something is fundamentally wrong and we should immediately exit.
```

Comment (natasha41575) [inline api/internal/localizer/localizer.go:111]: what is the linter complaining about?

File: api/internal/localizer/localizer.go [hunk 4]
```diff
@@ -75,9 +78,55 @@ func (lc *Localizer) Localize() error {
...
+			log.Fatalf("rel path error for 2 navigable roots %q, %q: %s", lc.ldr.Root(), ldr.Root(), err.Error())
```
Comment (natasha41575) [inline api/internal/localizer/localizer.go:112]:
```suggestion
			log.Fatalf("could not determine relative path between %q and %q: %s", lc.ldr.Root(), ldr.Root(), err.Error())
```
I understand that we don't expect to hit this, but if somehow this error ever does get surfaced to a user, it is currently very cryptic (even for me).

File: api/internal/localizer/localizer_test.go [hunk 5]
```diff
@@ -113,5 +160,161 @@ resources:
 	addFiles(t, fSysExpected, "/dst/a", map[string]string{
 		"kustomization.yaml": kustomization["Kustomization"],
 	})
-	require.Equal(t, fSysExpected, fSys)
+	checkFSys(t, fSysExpected, fSys)
+}
+
+func TestLocalizeDirInTarget(t *testing.T) {
+	for name, files := range map[string]map[string]string{
```
Comment (natasha41575) [inline api/internal/localizer/localizer_test.go:167]: I think when we get to doing `map[string]map[string]string`, it's getting a bit too complex. Suggest instead using this structure:

```
func TestLocalizeDirInTarget(t *testing.T) {
	type testCase struct {
		name  string
		files map[string]string
	}

	for _, tc := range []testCase{
		{
			name: "multi_nested_child",
			files: map[string]string{"kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1beta1
components:
- delta/epsilon
kind: Kustomization
`,
			},
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			...
		})
	}
}
```

File: api/internal/localizer/util.go [hunk 6]
```diff
@@ -14,6 +14,12 @@ import (
 	"sigs.k8s.io/kustomize/kyaml/filesys"
 )
 
+// DstPrefix prefixes the target and ref, if target is remote, in the default localize destination directory name
+const DstPrefix = "localized"
+
+// LocalizeDir is the name of the localize directories used to store remote content in the localize destination
+const LocalizeDir = "localized-files"
```
Comment (natasha41575) [inline api/internal/localizer/util.go:21]: nit: combine consts

```
const (
  // DstPrefix prefixes the target and ref, if target is remote, in the default localize destination directory name
  DstPrefix = "localized"

  // LocalizeDir is the name of the localize directories used to store remote content in the localize destination
  LocalizeDir = "localized-files"
)
```

File: api/internal/localizer/localizer_test.go [hunk 7]
```diff
@@ -113,5 +160,161 @@ resources:
...
+kind: Kustomization
+`,
+		"a/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+patches:
+- patch: |-
+    - op: replace
+      path: /spec/template/spec/containers/0/name
+      value: my-nginx
+replacements:
```
Comment (natasha41575) [inline api/internal/localizer/localizer_test.go:288]: Is there any particular reason to add patches and replacements and replicas to these component files? I had a similar question for some of the tests in https://github.com/kubernetes-sigs/kustomize/pull/4865 as well.

I think unless we are specifically testing a particular field, we should try to keep the test kustomization files minimal, maybe just a namePrefix or something. Otherwise it adds noise to the test.

File: api/internal/localizer/localizer.go (hunk 3 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:111]: Understood. Thanks for pointing this out!

Comment (annasong20) [reply api/internal/localizer/localizer.go:111]: If `log.Fatal` is hit, the `defer` statement earlier in this function isn't run. 

I think this is a sound check, but in our case, we only call `log.Fatal` if we hit an unexpected situation in which our existing assumptions are overthrown. In such a case, I think it's valid to immediately exit without trying to clean state? Open to other options as well.

File: api/internal/localizer/localizer_test.go (hunk 7 above)
Comment (annasong20) [reply api/internal/localizer/localizer_test.go:288]: Fixed.

I was trying to test that different fields in a `Component` are all copied as an extra precaution. I see your point about the clutter. The extra fields aren't strictly necessary because the logic used to process `Component`s is the same as that used for `Kustomization`s, and we've tested that already.

File: api/internal/localizer/localizer.go [hunk 8]
```diff
@@ -127,3 +134,43 @@ func (lc *Localizer) localizeFile(path string) (string, error) {
...
+	defer func() { _ = ldr.Cleanup() }()
+
+	var locPath string
+	if repo := ldr.Repo(); repo != "" {
+		// TODO(annasong): check if able to add localize directory
+		locPath = locRootPath(path, filesys.ConfirmedDir(repo), filesys.ConfirmedDir(ldr.Root()))
+	} else {
+		locPath, err = filepath.Rel(lc.ldr.Root(), ldr.Root())
+		if err != nil {
+			// This should never occur, but if hit, something is fundamentally wrong and we should immediately exit.
```
Comment (yuwenma) [inline api/internal/localizer/localizer.go:153]: I think the main difference between using `log.Fatalf` and `return nil, err` is that the latter actually gives users more info to triage and fix (from the layered wrapped error messages). Do you mind further explaining the benefits of using `log.Fatalf` here? If it's not that obvious, maybe we shall return the error.  And if we return error, maybe we do not need to skip the gocritic linter (which is another bonus)?

**Review by yuwenma — APPROVED** [review]
two minor comments. Otherwise lgtm

File: api/internal/localizer/localizer.go [hunk 9]
```diff
@@ -127,3 +134,43 @@ func (lc *Localizer) localizeFile(path string) (string, error) {
...
 	return locPath, nil
 }
+
+// localizeDir localizes root path and returns the localized path
+func (lc *Localizer) localizeDir(path string) (string, error) {
+	ldr, err := lc.ldr.New(path)
+	if err != nil {
+		return "", errors.Wrap(err)
+	}
+	defer func() { _ = ldr.Cleanup() }()
```
[the remark this answers is not in the document]
Comment (annasong20) [reply api/internal/localizer/localizer.go:144]: Yes. If I don't handle the error returned by `ldr.Cleanup()`, the `linter` complains that `

Error return value of `ldr.Cleanup` is not checked (errcheck)`

File: api/internal/localizer/localizer.go (hunk 8 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:153]: My initial reasoning for using `log.Fatal` here instead of `error` is to differentiate "impossible" errors - errors that should never be thrown - from normal errors. A `filepath.Rel` error would contradict our assumptions about `fileloader` checking for legal, navigable paths. If this error is thrown, there is a bug in our code that we should catch during testing or review. We should fail fast and there shouldn't be a way to silence/bypass this failure.

I believe there are similar use cases in kustomize code. For example, https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/filesys/fsondisk.go#L69

Down to change it to return the error if that's the better practice. Happy to defer to @yuwenma to make the call.

Comment (yuwenma) [reply api/internal/localizer/localizer.go:153]: That's fair.

File: api/internal/localizer/localizer_test.go [hunk 10]
```diff
@@ -299,3 +299,171 @@ patches:
...
+`,
+				"delta/epsilon/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1alpha1
+buildMetadata:
+- managedByLabel
+kind: Component
+`,
+			},
+		},
+		{
+			name: "localize_directory_name_when_absent",
```
Comment (KnVerey) [inline api/internal/localizer/localizer_test.go:379]: what does "absent" mean in this context? It looks like the name is "localized-files", but without much familiarity with the localize code I'm not sure why that is special.

File: api/internal/localizer/localizer_test.go [hunk 11]
```diff
@@ -299,3 +299,171 @@ patches:
...
+		})
+	}
+}
+
+func TestLocalizeDirCleanedSibling(t *testing.T) {
+	fSys := makeMemoryFs(t)
+	kustAndComponents := map[string]string{
+		"beta/gamma/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1beta1
+components:
+- delta/../../../../a/b/../../alpha/beta/sibling
```
Comment (KnVerey) [inline api/internal/localizer/localizer_test.go:414]: Is the point that this is expressed in a ridiculous way that looks like it is trying to escape, but once we clean it, it turns out to be a legal path, so it works?

File: api/internal/localizer/localizer.go [hunk 12]
```diff
@@ -127,3 +134,43 @@ func (lc *Localizer) localizeFile(path string) (string, error) {
...
+
+	var locPath string
+	if repo := ldr.Repo(); repo != "" {
+		// TODO(annasong): check if able to add localize directory
```
Comment (KnVerey) [inline api/internal/localizer/localizer.go:148]: What does this mean? I'm having trouble understanding this one

File: api/internal/localizer/localizer.go [hunk 13]
```diff
@@ -127,3 +134,43 @@ func (lc *Localizer) localizeFile(path string) (string, error) {
...
+		locPath = locRootPath(path, filesys.ConfirmedDir(repo), filesys.ConfirmedDir(ldr.Root()))
```
Comment (KnVerey) [inline api/internal/localizer/localizer.go:149]: IIUC this branch gets hit when the path was remote, but has already been cloned to a tempdir somewhere. So we do not expect the directory to be relative to the localizer's root (as in the other branch), and if we dig through the code we know it was previously made a `ConfirmedDir`. Still, I'm hesitant to simply convert the strings to ConfirmedDir like this without actually doing the confirmation--it means we're making assumptions that the current behaviour will be maintained. Is it even important to this method that the dir has been confirmed?

File: api/internal/localizer/localizer.go [hunk 14]
```diff
@@ -127,3 +134,43 @@ func (lc *Localizer) localizeFile(path string) (string, error) {
...
+	if err = lc.fSys.MkdirAll(newDst); err != nil {
+		return "", errors.WrapPrefixf(err, "unable to create root %q in localize destination", path)
+	}
+	err = (&Localizer{
+		fSys:      lc.fSys,
+		validator: lc.validator,
+		rFactory:  lc.rFactory,
+		pLdr:      lc.pLdr,
+		ldr:       ldr,
+		dst:       filesys.ConfirmedDir(newDst),
```
Comment (KnVerey) [inline api/internal/localizer/localizer.go:170]: similar comment here: should we be actually confirming this dir rather only declaring it to have been confirmed?

File: api/internal/localizer/localizer.go (hunk 8 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:153]: The counterargument is more theoretical than anything. The distinction between `log.Fatal` and `log.Panic` is that `log.Panic` allows for recovery and `defer` statements to run. In our use case, if this `log.Fatal` is hit, core assumptions have been violated, and we should exit immediately, no exceptions. We don't want to recover or give a chance for `defer` statements to further corrupt the existing state. The argument against needing to call `defer` statements to cleanup `localize` state is that because we only call `log.Fatal` under "impossible" cases, we should catch these cases during testing and shouldn't hit these cases during normal execution.

Given that this argument is purely theoretical, I can change instances of `log.Fatal` to `log.Panic` if you'd prefer, @KnVerey

Comment (KnVerey) [reply api/internal/localizer/localizer.go:153]: I see what you're saying about these being things that should never happen in normal execution, but in short, I prefer panicking in such cases (if there's no way around them to begin with!). I don't think there's much risk of us letting in code that recovers arbitrary panics (the only recovery I see is inside the go-yaml fork fwiw), and I feel like the cleanup code we typically have deferred should be safe to run regardless, e.g. I'd see cleaning up temporary state we've created on disk as desirable. Two other things to consider: first, if you allow the test execution to recover, you can have a nicer debugging experience where test cases other than the triggering case can continue. Second, if the problem _does_ end up happening in normal execution, the panic output is maximally clear that this is not the end user's fault and should be reported.

Comment (annasong20) [reply api/internal/localizer/localizer.go:153]: Thanks for the detailed explanation!! Got it. Will change.

File: api/internal/localizer/localizer_test.go (hunk 10 above)
Comment (annasong20) [reply api/internal/localizer/localizer_test.go:379]: It's referring to an absent remote reference. This test case is testing that a localize directory name - name of directory that we create to store localized files - can be used in the absence of a remote reference. In such a case, `kustomize localize` doesn't need to create a localize directory, and should allow the `target` directory to contain directories with the localize directory name.

I'll rename the test case to make it more comprehensible

File: api/internal/localizer/localizer.go (hunk 12 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:148]: To add a localize directory here, we should check if there's a directory at this location in `scope` that shares the localize directory name. In such a case, we should return an error to avoid confounding directories.

File: api/internal/localizer/localizer.go (hunk 13 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:149]: Yes, `locRootPath` needs the `repo` to be a confirmed directory because it wants to create a localized path without symlinks. 

I can actually call `ConfirmDir`. 

Not for this PR, but what do you think about changing `ifc.Loader`'s `Root()` method to return a `filesys.ConfirmedDir` instead? It shouldn't break implementations of `ifc.Loader` for testing because filesystems in memory can also have `filesys.ConfirmedDir`, right?

File: api/internal/localizer/localizer.go (hunk 14 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:170]: Will do.

File: api/internal/localizer/localizer.go (hunk 13 above)
Comment (KnVerey) [reply api/internal/localizer/localizer.go:149]: That idea makes sense to me, and offhand I _feel_ like it should work fine.

File: api/internal/localizer/localizer.go (hunk 12 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:148]: I forgot to add that we don't throw an error for an unreferenced directory that shares the localize directory name. It's only a problem when the local directory is referenced and copied by `localize` AND there's a remote reference.

File: api/internal/localizer/localizer.go (hunk 14 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:185]: I don't think we need to call `filesys.ConfirmDir()` in this case because the previous `dst` is a `filesys.ConfirmedDir`, `toDst` is clean, and we call `MkdirAll` to create `toDst`, which, as a result, cannot contain symlinks. 

I changed all other instances of the conversion that actually assumed `Root()` and `Repo()` returned `ConfirmedDir`s, though.

File: api/internal/localizer/locloader_test.go [hunk 15]
```diff
@@ -127,65 +127,6 @@ func TestNewMultiple(t *testing.T) {
...
-
-			req.Equal("a/b/c/d/e", args.Target.String())
-			req.Equal("a/b/c", args.Scope.String())
-			req.Equal(test.wd+"/"+test.newDir, args.NewDir.String())
-			// memory file system can only find paths rooted at current node
-			req.True(fSys.Exists(test.newDir))
-		})
-	}
-}
-
```
Comment (annasong20) [inline api/internal/localizer/locloader_test.go:188]: I deleted these tests that use non-root`fsnode`s as the working directory because `fsnode`s only understand relative paths. Before we called `ConfirmDir()` on the established root and repo directories, which are absolute paths, `ConfirmDir()` was only called on relative paths in `fileLoader` in these test cases.

I will cover working directory behavior during integration testing in a future PR.

File: api/internal/localizer/localizer.go (hunk 14 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:185]: There shouldn't be any symlinks because `newDst` must be in the localize destination directory, which `LocLoader` created. 

I like the suggestion, but I know it's a redundant check. I'm thinking I'll change the structure of the `Localizer` code so that the destination directory is never exposed. Then, we can be sure that it doesn't contain any symlinks.

Comment (annasong20) [reply api/internal/localizer/localizer.go:205]: Should be addressed.

Comment (annasong20) [reply api/internal/localizer/localizer.go]: @KnVerey ready for review

Following our very helpful discussion today, I realized that `dst` (contrary to some other values) doesn't actually need to be a `filesys.ConfirmedDir`. I changed the `filesys.ConfirmedDir` to `string`.

File: api/internal/localizer/localizer.go (hunk 12 above)
Comment (annasong20) [reply api/internal/localizer/localizer.go:177]: Will simplify this behavior, according to our discussion today.

File: api/internal/localizer/locloader_test.go (hunk 15 above)
Comment (annasong20) [reply api/internal/localizer/locloader_test.go]: No longer applicable.

**Comment by annasong20** [comment]
@KnVerey Sorry for all the changes that make the PR history hard to follow. I addressed all of your feedback pertaining to the code to localize directories. There are a few other instances of existing code that I need to apply your feedback to, but I want to limit the scope of this PR to make it easier to review. I will address the other instances in a separate PR. 

This PR is now ready for review.

**Comment by annasong20** [comment]
Rebased.
